### PR TITLE
Ensure lazy evaluation for probs and logits

### PR DESCRIPTION
**Pull request by neerajprad**
This makes a few modifications to ensure that `logits` and `probs` are not accessed by a method unless it needs to. e.g. accessing `probs` on a class that was initialized with `logits` only to get the batch shape or event shape. This affects Categorical, OneHotCategorical, Bernoulli and Multinomial.

**Testing:** Added some tests to prevent future regression:
 - `initializing` a distribution with `probs` and calling `sample`, `enumerate_support`, `batch_shape` and `event_shape` should not initialize `logits`.
 - Conversely, `initializing` a distribution with `logits` and calling `log_prob`, `enumerate_support`, `batch_shape` and `event_shape` should not initialize `probs`.

### CODE REVIEW

**Comment by neerajprad** [comment]
@alicanb, @fritzo  - Does this look fine to be opened upstream as PR in pytorch?

**Comment by alicanb** [comment]
If we make multinomials properties lazy, will there be 2 copies `self.probs` and `self._categorical.probs`?

**Comment by neerajprad** [comment]
> If we make multinomials properties lazy, will there be 2 copies self.probs and self._categorical.probs?

Making it lazy is to ensure that both are not initialized on construction. When you need one of `logits` or `probs`, the underlying categorical implementation will be called and the local instance attribute will be updated to point to that - there should only be a single copy though.

File: torch/distributions/bernoulli.py [hunk 1]
```diff
@@ -33,14 +33,16 @@ def __init__(self, probs=None, logits=None):
...
             raise ValueError("Either `probs` or `logits` must be specified, but not both.")
         if probs is not None:
+            is_scalar = isinstance(probs, Number)
             self.probs, = broadcast_all(probs)
         else:
+            is_scalar = isinstance(logits, Number)
             self.logits, = broadcast_all(logits)
-        probs_or_logits = probs if probs is not None else logits
-        if isinstance(probs_or_logits, Number):
+        self.probs_or_logits = self.probs if probs is not None else self.logits
```
Comment (fritzo) [inline torch/distributions/bernoulli.py]: :+1: to adding this field. Consider renaming to `self._prototype` or `self._template` or `self._param`?

File: torch/distributions/categorical.py [hunk 2]
```diff
@@ -44,13 +44,15 @@ def __init__(self, probs=None, logits=None):
             self.probs = probs / probs.sum(-1, keepdim=True)
         else:
             self.logits = logits - log_sum_exp(logits)
-        self._num_events = self.probs.size()[-1] if self.probs is not None else self.logits.size()[-1]
-        batch_shape = self.probs.size()[:-1] if probs is not None else self.logits.size()[:-1]
+        self.probs_or_logits = self.probs if probs is not None else self.logits
+        self._num_events = self.probs_or_logits.size()[-1]
+        self.param_shape = self.probs_or_logits.size()
```
Comment (fritzo) [inline torch/distributions/categorical.py]: If you rename `self.probs_or_logits` to `self._param` then this would become `self._param.shape`.

File: torch/distributions/multinomial.py [hunk 3]
```diff
@@ -47,19 +47,19 @@ def __init__(self, total_count=1, probs=None, logits=None):
...
+        batch_shape = self._categorical.batch_shape
+        event_shape = self._categorical.probs_or_logits.size()[-1:]
         super(Multinomial, self).__init__(batch_shape, event_shape)
 
     @constraints.dependent_property
     def support(self):
         return constraints.integer_interval(0, self.total_count)
 
-    @property
+    @lazy_property
```
Comment (fritzo) [inline torch/distributions/multinomial.py]: Would this be cheaper as simply `.property`?

**Comment by fritzo** [comment]
@alicanb can you ensure that these changes also make their way to your `Binomial` and `Geometric` PRs that have not yet merged?

File: torch/distributions/categorical.py (hunk 2 above)
Comment (fritzo) [reply torch/distributions/categorical.py]: Also, this should not be public imo.

**Comment by alicanb** [comment]
Sure, let's get this in before binomial, I can make appropriate changes afterwards.

File: torch/distributions/bernoulli.py (hunk 1 above)
Comment (neerajprad) [reply torch/distributions/bernoulli.py]: Sure, will rename to `self._param`. I get warnings that a private property is accessed from outside the class, but it should be fine since this isn't part of the API and only used internally within distributions.

Comment (fritzo) [reply torch/distributions/bernoulli.py]: Oh interesting, is that in your IDE? I guess those meaning of underscore both make sense (private to class -vs- not part of public API), and I hadn't thought of that before. In this case it seems cleaner to ensure it's not part of the public API.

What might make sense to add to the public API would be a `.new()` method and a `.param_shape` property (as you have added in this PR). Are there any attributes we use of `_param` other than `.shape` and `.new()`?

File: torch/distributions/multinomial.py (hunk 3 above)
Comment (neerajprad) [reply torch/distributions/multinomial.py]: Actually, both are the same for all practical purposes (ran a simple profiler), since the underlying implementation is lazy. Will simply revert to `@property`. cc. @alicanb

[the remark this answers is not in the document]
Comment (fritzo) [reply torch/distributions/multinomial.py]: there will be one copy that is pointed to from two places.

File: torch/distributions/bernoulli.py (hunk 1 above)
Comment (fritzo) [reply torch/distributions/bernoulli.py]: Adam said adding a `.new()` method is pretty common practice, so `.new()` + `.param_shape` (both public) seems pretty clean. WDYT?

Comment (fritzo) [reply torch/distributions/bernoulli.py]: If we add a `.new()` method here, we could send another PR that adds them in other distributions and simplifies a bunch of the `_kl_infinite` logic.

**Comment by neerajprad** [comment]
Opened as https://github.com/pytorch/pytorch/pull/4691.
